File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    answer = 0
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    alphabet += alphabet.lower()
    
    # str1 두 글자씩 끊어서 다중집합 만들기
    str1_array = []
    for i in range(len(str1) - 1):
        if (str1[i] in alphabet) and (str1[i + 1] in alphabet):
            str1_array.append((str1[i] + str1[i + 1]).upper())
            
    # str2 두 글자씩 끊어서 다중집합 만들기
    str2_array = []
    for i in range(len(str2) - 1):
        if (str2[i] in alphabet) and (str2[i + 1] in alphabet):
            str2_array.append((str2[i] + str2[i + 1]).upper())

    # 교집합 만들기
    AandB = []
    for item in str1_array:
        if item not in AandB:
            # 둘다 있으면 그 중 최소값
            if item in str2_array:
                AandB += [item for _ in range(min(str1_array.count(item), str2_array.count(item)))]
    
    # 합집합 만들기
    AorB = []
    for item in str1_array:
        if item not in AorB:
            if item in str2_array:
                AorB += [item for _ in range(max(str1_array.count(item), str2_array.count(item)))]
            else:
                AorB += [item for _ in range(str1_array.count(item))]
    for item in str2_array:
        if item not in AorB:
            if item in str1_array:
                AorB += [item for _ in range(max(str1_array.count(item), str2_array.count(item)))]
            else:
                AorB += [item for _ in range(str2_array.count(item))]

    if len(AandB) == 0 and len(AorB) == 0:
        return 65536
    return int((len(AandB) / len(AorB)) * 65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
// 문자를 두 글자씩 끊어서 다중 집합으로 만드는 함수
const convertStr = (str) => {
    str = str.toUpperCase();
    const ref = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    
    const result = [];
    for (let i = 0; i < str.length - 1; i++) {
        // 공백 체크
        if (str[i] && str[i + 1]) {
            // 숫자 체크
            if (isNaN(str[i]) && isNaN(str[i])) {
                // 특수문자 체크
                if (ref.includes(str[i]) && ref.includes(str[i + 1])) {
                    result.push(str[i] + str[i + 1]);
                }
            }
        }
    }
    
    return result;
}

function solution(str1, str2) {
    let answer = 0;
    
    const convertStr1 = convertStr(str1);
    const convertStr2 = convertStr(str2);
    
    if (!convertStr1.length && !convertStr2.length) answer = 1;
    else {
        const totalArr = [...new Set(convertStr1.concat(convertStr2))];
        const str1Map = new Map();
        convertStr1.forEach((str) => str1Map.set(str, str1Map.has(str) ? str1Map.get(str) + 1 : 1));
        const str2Map = new Map();
        convertStr2.forEach((str) => str2Map.set(str, str2Map.has(str) ? str2Map.get(str) + 1 : 1));
        
        // 교집합
        let intersection = 0;
        totalArr.forEach((item) => {
            if (str1Map.has(item) && str2Map.has(item)) intersection += Math.min(str1Map.get(item), str2Map.get(item));
        })
        
        let union = 0;
        totalArr.forEach((item) => {
            if (str1Map.has(item) && str2Map.has(item)) union += Math.max(str1Map.get(item), str2Map.get(item));
            else if (str1Map.has(item)) union += str1Map.get(item);
            else if (str2Map.has(item)) union += str2Map.get(item);
        })
        
        answer = intersection / union
    }
    
    return parseInt(answer * 65536);
}